### ISSUE TITLE
Unexpected behavior when `IndentationError` raise

### ISSUE THREAD

**Issue by F-park** [python/cpython#113079]
# Bug report

### Bug description:
In python 3.8
```pycon
>>>     0
  File "<stdin>", line 1
    0
    ^
IndentationError: unexpected indent
>>> import sys, traceback
>>> sys.excepthook = lambda *exc_info: traceback.print_exception(*exc_info)
>>>     0
  File "<stdin>", line 1
    0
    ^
IndentationError: unexpected indent
```

In python 3.9 ... 3.13.0a3#e7331365b488382d906ce6733ab1349ded49c928^
```pycon
>>>     0
  File "<stdin>", line 1
    0
IndentationError: unexpected indent
>>> import sys, traceback
>>> sys.excepthook = lambda *exc_info: traceback.print_exception(*exc_info)
>>>     0
  File "<stdin>", line 1
        0
    ^
IndentationError: unexpected indent
```

In python 3.13.0a3#e7331365b488382d906ce6733ab1349ded49c928 ... main
```pycon
>>>     0
  File "<stdin>", line 1
        0
    ^
IndentationError: unexpected indent
```

### CPython versions tested on:

3.9 ... 3.13.0a3#e7331365b488382d906ce6733ab1349ded49c928^

### Operating systems tested on:

Windows

**Comment by gaogaotiantian** [comment]
I'm a bit confused here. The exception looks normal to me. Could you post what you think is incorrect?

**Comment by cyx2015s** [comment]
The underlying up arrow mark points at different location between different versions.

**Comment by terryjreedy** [comment]
@pablogsal The issue is the '^' location after an IndentationError.  In 3.13.0a2 and current main, there is no caret.  Is this an intentional fix of it being misplaced?

**Comment by F-park** [comment]
> I'm a bit confused here. The exception looks normal to me. Could you post what you think is incorrect?

In python 3.8
```pycon
>>>     0
  File "<stdin>", line 1
    0
    ^
IndentationError: unexpected indent
```

In python 3.9+
```pycon
>>>     0
  File "<stdin>", line 1
    0
IndentationError: unexpected indent
```

I consider in python 3.9+, the correct error should be the below which print by `traceback.print_exception`
```pycon
>>>     0
  File "<stdin>", line 1
        0
    ^
IndentationError: unexpected indent
```

**Comment by gaogaotiantian** [comment]
The direct cause of this is the parser change in 3.9, which determines the column number of the actual error. In 3.8, the parser thinks the error is at the first character, which works fine with the carets. From 3.9, the parser believes the error is at the space before the first character, but the exception formatter will dedent the line with the syntax error, which leaves no space for carets and therefore no carets are printed. In repl I believe there are extra spaces so the carets are available again.

 I guess the question here is - what's the expected behavior?

```
  File "/home/gaogaotiantian/programs/mycpython/scrabble.py", line 3
     b = 2
    ^
IndentationError: unexpected indent
```

Is this better? Notice that the source line `b = 2` is indented by an extra space for the caret.

**Comment by gaogaotiantian** [comment]
The current behavior is intentional from https://github.com/python/cpython/commit/15bc9ab301d73f20bff47a12ef05326feb40f797 and maybe @gvanrossum can share thoughts on this? Do you believe suppressing the caret for the `IndentationError` is the preferred behavior?

**Comment by F-park** [comment]
The bug fixed in e7331365b488382d906ce6733ab1349ded49c928 , but it not backport to 3.9 ... 3.12.

**Comment by gvanrossum** [comment]
Assuming the fix won't be backported (too fiddly, not very important), do you still need my opinion? I have to admit that I don't recall the details around 15bc9ab, so it would take considerable time for me to reconstruct what's going on here and form another opinion...

**Comment by gaogaotiantian** [comment]
It's not "fixed". It happens to not appear on REPL, but the same issue happens with scripts.

It's actually a pretty simple choice (to save you from recollections):

```python
    print("line with unexpected indent")
```

Do you think we should do:
```
  File "/home/gaogaotiantian/programs/mycpython/example.py", line 1
    print("hello")
IndentationError: unexpected indent
```
Or
```
  File "/home/gaogaotiantian/programs/mycpython/example.py", line 1
    print("hello")
   ^
IndentationError: unexpected indent
```

We could have other possible indentations for the actual line in the traceback, the core question is - do you think there should be a caret pointing to the space before the `print` (or maybe to the `print`)?

**Comment by gvanrossum** [comment]
Comparing various indentation errors in the REPL:
```
(.venv) ~/cpython$ python3.13
Python 3.13.0a4 (v3.13.0a4:9d34f60783, Feb 15 2024, 10:52:45) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> def f():
... pass
  File "<stdin>", line 2
    pass
    ^^^^
IndentationError: expected an indented block after function definition on line 1
>>> def f():
...     pass
...   pass
  File "<stdin>", line 3
    pass
        ^
IndentationError: unindent does not match any outer indentation level
>>> def f():
...     pass
...       pass
  File "<stdin>", line 3
    pass
IndentationError: unexpected indent
>>> def f():
...     pass
... pass
  File "<stdin>", line 3
    pass
    ^^^^
SyntaxError: invalid syntax
>>> 
```
It looks like the "unexpected indent" one is the odd one out that doesn't print the caret. So it looks like a mistake.

But probably we should ask @pablogsal, who I'm sure will have an opinion. (Hi, Pablo!)

**Comment by gaogaotiantian** [comment]
Yes, the reason "unexpected indent" is the odd one is the caret is **before** the actual statement. The statement is dedented in the traceback so the "supposed position" of the caret is negative, which will be suppressed.

If we want to change the behavior, we need to decide where we should put the caret - from the bug picture, there are extra spaces for the caret, but the traceback formatter probably does not know that. The indentation is inserted by multiple layers and when traceback formatter calculates the position of the caret, it believes there's no position for it.

**Comment by gvanrossum** [comment]
It should be put at the start of the line, I.e. below the first token. I have no idea where that code is any more, sorry. (It’s always been fiddly, and has probably only gotten more so with various improvements over the past 5 years.)

**Comment by gaogaotiantian** [comment]
No worries, I found the code that produces this error and kind of figured out what happened, so no code tracking work for you. However, the new parser (we changed to in 3.9) believes the parsing error is before the first token - on the last space before the character, that's why the behavior changed in 3.9. If we want to put the caret on the first token, we need to set special rule for this, overriding what the parser generates.

**Comment by gvanrossum** [comment]
But the indents are generated by the lexer, which can do as it pleases.

**Comment by gaogaotiantian** [comment]
So what you meant is we should change the output of the compiler to locate the indentation error at the first character of the token, instead of trying to locate the caret after the indentation error located at negative index.

**Comment by gvanrossum** [comment]
It depends. But I fear that some intermediate layers would interpret a negative value as "absent", so that might be a problem. If you can fix it in the display layer and it doesn't cause other problems, that would be fine too of course.

**Comment by gaogaotiantian** [comment]
Sorry I was describing the issue too vaguely. with
```python
    print("hello")
````
The compiler believes the error is at column 3(0-index), the space right before `print`. For all the intermediate layers, that worked fine because it is a valid positive value that exists in the file.

However, when `traceback` is printing the error, it will dedent the line first - presumably to avoid print unnecessary spaces before the statement. So instead of
```python
    print("hello")
```
It will print
```python
print("hello")
```

Then `traceback` recalculates where the caret should be based on the information given by the compiler. The error position was originally at column 3, but as the line is dedented by 4, now it should be at column -1, and that's why it's not printed.

We have a couple of solutions here:
1. Change the compiler, so the error position is at column 4(the start character `p` of `print`) instead of 3. This might involve some hack to the parser and I don't know if we want to do that.
2. Change the `traceback` formatter. As the only possible error case (that I can think of) as such is indentation error (which dedentation shifts the actual error position to negative), we can simply output caret at index 0, if it's supposed to be at index -1.
3. Keep the relative relation of the error position (column 3) and the statement (starting at column 4), somehow right shift the statement line by 1 (which will break the alignment among all the statements), or try to display the caret at the column before the line (because currently all the traceback prints have natural indent which can try to steal).

**Comment by gvanrossum** [comment]
Either (1) or (2) will work.

**Comment by pablogsal** [comment]
> But probably we should ask @pablogsal, who I'm sure will have an opinion. (Hi, Pablo!)

👋  Thanks for checking with me!  I have catch up with the discussion. I am going to investigate what solution it's more cleaner here. My guess is that this should be fixed in the lexer to avoid having to sync many different pieces downstream but it's possible that the lexer will require extra state to deal with this, in which case I will investigate if we can easily fix the display.

I am a bit hesitant to just fix this in the traceback code because there are other possible consumers of the column information (formatting tools and the like) and I don't want to have a custom rule at the end when we fix a mistake that should have been fixed before. But I also don't want to overcomplicate the lexer.

I will take a look and make a PR or report back with what I find.
